Add RedshiftDataHook

john-jac opened the pull request:
Airflow users wishing to trigger Amazon Redshift queries today have to use a Postgres connection.  While effective, this has the limitation that the Redshift cluster must expose a Postgres endpoint that is accessible to the Airflow environment.

As an alternative, the AWS Boto3 library exposes a client called [redshift-data](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/redshift-data.html) that allows users to query and retrieve data via the AWS API and avoid the necessity of a Postgres connection.

This PR adds a RedshiftDataHook to the existing Redshift provider, along with tests and an example, which wraps the "redshift-data" Boto3 client and allows users to run Redshift queries without a Postgres connection.

In the future, this capability could be added to any existing Redshift transfer operator, with an option to auto-detect whether to use Postgres or Boto3 based upon the connection type.

ephraimbuddy commented on airflow/providers/amazon/aws/example_dags/example_redshift_data_execute_sql.py:
```diff
@@ -0,0 +1,118 @@
+
+@task(task_id="execute_query")
+def execute_query_fn():
+    """This is a python callback that executes a Redshift query"""
```
```suggestion
    """This is a python decorator task that executes a Redshift query"""
```

ephraimbuddy commented on airflow/providers/amazon/aws/example_dags/example_redshift_data_execute_sql.py:
```diff
@@ -0,0 +1,118 @@
+
+@task(task_id="wait_for_results")
+def wait_for_results_fn(id):
+    """This is a python callback that executes a Redshift query"""
```
```suggestion
    """This is a python decorator task that executes a Redshift query"""
```

ephraimbuddy commented on airflow/providers/amazon/aws/example_dags/example_redshift_data_execute_sql.py:
```diff
@@ -0,0 +1,118 @@
+    # Using a task-decorated function to request the list of tables in a Redshift cluster
+    redshift_query = execute_query_fn()
+
+    # Using a task-decorated function to wait for the Redshift query above
```
Since you are using the Taskflow API, I suggest you also use `@dag` decorator

ephraimbuddy reviewed (changes requested):
Why can't we use RedshiftSQLHook or RedshiftHook? Why do we need a new hook? It's not clear to me.

I suggest we add these methods to the existing RedshiftHook and create operator(s) that would do what we want

john-jac commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,127 @@
+            elif status == 'FAILED':
+                raise ValueError(f"RedshiftDataHook.describe_statement {status}")
+            elif status == 'ABORTED':
+                raise ValueError(f"Query {status}")
```
RedshiftDataHook wraps execute_statement, which executes a single SQL statement.  There is a separate call, [batch_execute_statement](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/redshift-data.html#RedshiftDataAPIService.Client.batch_execute_statement), that will run multiple SQL statements.  This functionality could be added in the future.

ephraimbuddy commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,127 @@
+        aws_conn_id: str = 'aws_default',
+        cluster_identifier: str,
+        database: str,
+        sql: str,
```
Can we make this to take a `list` too, that way we will support running multiple SQL statements at once?

john-jac replied:
Multiple SQL statements is a separate boto3 call, [batch_execute_statement](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/redshift-data.html#RedshiftDataAPIService.Client.batch_execute_statement), which could be added in a future release.

ephraimbuddy commented on airflow/providers/amazon/aws/example_dags/example_redshift_data_execute_sql.py:
```diff
@@ -0,0 +1,83 @@
+    catchup=False,
+)
+def example_redshift_data():
+    @task(task_id="output_results")
+    def output_results_fn(id):
+        """This is a python decorator task that returns a Redshift query"""
+        hook = RedshiftDataHook()
+
+        resp = hook.get_statement_result(
+            id=id,
+        )
+        print(resp)
+        return resp
```
Since this is necessary, we should create an operator for it. No strong opinion but it seems necessary.

john-jac replied:
For this first release the aim is parity with the Postgres operator, which I don't believe returns a full set of results (but I could be mistaken).  Certainly for a future release we could break this out, and in fact it would be good to add a sensor as well so that the SQL statement can be performed asynchronously.

ephraimbuddy commented on airflow/providers/amazon/aws/hooks/redshift_data.py:
```diff
@@ -0,0 +1,152 @@
+                WithEvent=with_event,
+                SecretArn=secret_arn,
+                StatementName=statement_name,
+            )
```
Since the `parameters` arg is not required, we could pass `None` to the execute_statement and it would be ignored.

john-jac replied:
In testing, passing None caused a failure in boto3 as invalid parameters.  That's why I had to break it out into two different calls.

vincbeck replied:
See related: https://github.com/boto/boto3/issues/331

ephraimbuddy commented on airflow/providers/amazon/aws/hooks/redshift_data.py:
```diff
@@ -0,0 +1,152 @@
+        response = self.get_conn().describe_statement(
+            Id=id,
+        )
+        return response['Status']
```
I would prefer we return the whole response here rather than just the status. Someone else might need to get more information other than `status`

john-jac replied:
Done

ephraimbuddy commented on airflow/providers/amazon/aws/hooks/redshift_data.py:
```diff
@@ -0,0 +1,152 @@
+        response = self.get_conn().cancel_statement(
+            Id=id,
+        )
+        return response['Status'] if response['Status'] else None
```
```suggestion
        return response['Status']
```
Do you suspect that it could return None?

john-jac replied:
Done

ephraimbuddy commented on airflow/providers/amazon/aws/hooks/redshift_data.py:
```diff
@@ -0,0 +1,152 @@
+        :type with_event: bool
+
+        """
+        """only provide parameter argument if it is valid"""
```
```suggestion
        # only provide parameter argument if it is valid
```
We may not need this comment though

john-jac replied:
Done

ephraimbuddy replied:
Looks not done yet

potiuk replied:
Yeah. likely missed during rebase.

ephraimbuddy commented on tests/providers/amazon/aws/hooks/test_redshift_data.py:
```diff
@@ -0,0 +1,76 @@
+
+from airflow.providers.amazon.aws.hooks.redshift_data import RedshiftDataHook
+
+
```
For this test, check that we can be able to connect and that we are actually calling the API we connected to. The current tests doesn't guarantee that we are hitting that API

john-jac replied:
Working on adding a system test and improving the unit test...will update PR when ready.  Thanks!

vincbeck replied:
I am planning to improve unit tests and add system tests. I just realized [moto](https://github.com/spulec/moto) does not cover redshift-data. I am planning to add it in moto, and then I'll be able to move forward this PR

dstandish replied:
this is a very simple hook / operator and i think using moto may be overcomplicating it. we can just write simple unit tests and conventional mocking techniques.  there are very few moving parts. basically we should just need to check that 1. execute_statement should be called with the operator params and 2. execution statuses should be handled properly, 3 nowait behavior, 4 xcom behavior. if moto is still holding you back i would not worry about it.

ephraimbuddy reviewed (dismissed):
Another thing that's required here is a system test. That would help us to verify that the example works as expected.

See https://github.com/apache/airflow/blob/main/tests/providers/amazon/aws/operators/test_ecs_system.py for an example system test. 

Here too is more information about a system test: https://github.com/apache/airflow/blob/main/TESTING.rst#airflow-system-tests

josh-fell commented on airflow/providers/amazon/aws/example_dags/example_redshift_data_execute_sql.py:
```diff
@@ -0,0 +1,83 @@
+    # Using a task-decorated function to output the list of tables in a Redshift cluster
+    redshift_output = output_results_fn("{{task_instance.xcom_pull('redshift_query', key='return_value')}}")
+
+    redshift_query >> redshift_output
```
```suggestion
    # Using a task-decorated function to output the list of tables in a Redshift cluster
    redshift_output = output_results_fn(redshift_query.output)
```
This can be simplified to use `XComArgs` to access the return value from the `RedshiftDataOperator` task directly without Jinja using the `.output` property. Also, the task dependencies would be handled implicitly as well.

Similar updates were made to the existing example DAGs as part of [this issue](https://github.com/apache/airflow/issues/10285). We should continue this approach when possible with new example DAGs.

john-jac replied:
Done

josh-fell commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,127 @@
+
+    template_fields = ('sql',)
+    template_ext = ('.sql',)
+
```
```suggestion
template_fields_renderers = {'sql': 'sql'}
```
WDYT about adding the field renderer attribute so the SQL is nicely readable in the UI?

john-jac replied:
Done

josh-fell commented on docs/apache-airflow-providers-amazon/operators/redshift_data.rst:
```diff
@@ -0,0 +1,51 @@
+Purpose
+"""""""
+
+This is a basic example dag for using :class:`RedshiftDataOperator <airflow.providers.amazon.aws.operators.redshift_data>`
```
```suggestion
This is a basic example DAG for using :class:`RedshiftDataOperator <airflow.providers.amazon.aws.operators.redshift_data>`
```

john-jac replied:
Done

josh-fell commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,127 @@
+
+    :param sql: the sql code to be executed
+    :type sql: Can receive a str representing a sql statement,
+        or an iterable of str (sql statements)
```
```suggestion
    :param sql: the sql code to be executed
    :type sql: str
```
Should update this if the the support for a list of SQL won't be supported with this PR.

john-jac replied:
Done

josh-fell commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,127 @@
+    :type autocommit: bool
+    """
+
+    template_fields = ('sql',)
```
Any other fields to add here that users might want to have flexibility to dynamically generate a value for? Maybe `parameters`, `cluster_identifier`, `database`, and/or `db_user`. The latter 3 could maybe be added to a Connection Extra and accessed with Jinja now that the `Connection` object is accessible in the template context like `"{{ conn.conn_id.extra_dejson.<attr> }}"` when calling the operator. No strong opinions though just something to think about.

potiuk replied:
Yeah. The more  template_fields the better. It has pretty much no side effects but makes your operators much more flxible and allow to build much more sophissticated cases.

josh-fell commented on docs/apache-airflow-providers-amazon/operators/redshift_data.rst:
```diff
@@ -0,0 +1,51 @@
+
+Use the :class:`RedshiftDataOperator <airflow.providers.amazon.aws.operators.redshift_data>` to execute
+statements against an Amazon Redshift cluster.
+
```
WDYT about adding some more context here about how this differs from the `RedshiftSQLOperator`? I suspect @ephraimbuddy and I won't be the only ones wondering.

john-jac replied:
Done

josh-fell commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,127 @@
+from airflow.providers.amazon.aws.hooks.redshift_data import RedshiftDataHook
+
+
+class RedshiftDataOperator(BaseOperator):
```
Since the `RedshiftDataHook.cancel_statement()` method is being introduced, WDYT about adding the `on_kill()` method override for this operator which run the `cancel_statement()` method? This would give users an opportunity to kill the query from the UI if needed.

potiuk replied:
I second @josh-fell here. Adding "on_kill" is often missing in the operators and that has some bad effects for our users.

john-jac replied:
OK sounds good.  @vincbeck is working on the outstanding issues including this one.

josh-fell commented on docs/apache-airflow-providers-amazon/operators/redshift_data.rst:
```diff
@@ -0,0 +1,52 @@
+Use the :class:`RedshiftDataOperator <airflow.providers.amazon.aws.operators.redshift_data>` to execute
+statements against an Amazon Redshift cluster.
+
+This differs from RedshiftSQLOperator in that it allows users to query and retrieve data via the AWS API and avoid the necessity of a Postgres connection.
```
Is this not necessarily true after version 2.4.0 of the Amazon provider is released? Looks like the Postgres dependency is being removed as part of #18844.

john-jac replied:
> Is this not necessarily true after version 2.4.0 of the Amazon provider is released? Looks like the Postgres dependency is being removed as part of #18844.

This is referring to the Redshift requirement that a Postgres URL and port is open to Airflow.  With boto3 redshift-data, no such access need exist

josh-fell replied:
Cool, just wanted to confirm. Thanks!

ephraimbuddy commented on airflow/providers/amazon/aws/hooks/redshift_data.py:
```diff
@@ -0,0 +1,152 @@
+                WithEvent=with_event,
+                SecretArn=secret_arn,
+                StatementName=statement_name,
+            )
```
```suggestion
        response = self.get_conn().execute_statement(
                ClusterIdentifier=cluster_identifier,
                Database=database,
                Sql=sql,
                DbUser=db_user,
                WithEvent=with_event,
                SecretArn=secret_arn,
                StatementName=statement_name,
                Parameters=parameters,
            )
```

Since `parameters` is not required https://docs.aws.amazon.com/redshift-data/latest/APIReference/API_ExecuteStatement.html

john-jac replied:
boto3 behaved differently when passed "None" as Parameters vs not passing Parameters at all, hence the need fot the if-else

dstandish commented on airflow/providers/amazon/aws/hooks/redshift_data.py:
```diff
@@ -0,0 +1,126 @@
+        kwargs["client_type"] = "redshift-data"
+        super().__init__(*args, **kwargs)
+
+    def execute_statement(
```
I would suggest that you simplify this hook to remove the methods which are essentially an alias for the boto3 client methods. Such as this one.

Instead, just use the underlying client directly in your operator, and add the library https://pypi.org/project/mypy-boto3-redshift-data/ and type hint the client object so that users still get nice autocomplete.

We are doing the exact same thing here: https://github.com/apache/airflow/pull/20642/files#diff-07892cfb7f2a919afec9185e36b417d75d5322d243096667ca1cf05c506d9533R51-R59

Implementing this kind of "alias" method doesn't really add any value but does introduce a maintenance burden.

vincbeck replied:
Done

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+        self,
+        database: str,
+        sql: str,
+        cluster_identifier: Optional[str] = "",
```
we generally use `None` for default value for optionals

vincbeck replied:
Fair enough. Done

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+        secret_arn: Optional[str] = "",
+        statement_name: Optional[str] = "",
+        with_event: bool = False,
+        timeout: int = 300,
```
tasks already have an `execution_timeout`

why add another one?

vincbeck replied:
That's a good point. Thanks for bringing it up. I removed it

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+    :param secret_arn: the name or ARN of the secret that enables db access
+    :param statement_name: the name of the SQL statement
+    :param with_event: indicates whether to send an event to EventBridge
+    :param timeout: how long in seconds to wait for a response, if 0 don't wait
```
very much not obvious that `timeout` means, also, 'sync'

i would make this boolean (since there's already a timout out in base op you can chop this), and call it `nowait` or something

vincbeck replied:
I agree. Though, I renamed it to `wait`, I dont really like having a negation in a boolean variable name

dstandish replied:
yeah that's fine with me, i don't really like it either :) 
just suggested it cus that's what emerged as consensus in another pr

dstandish replied:
though `wait` alone feels a little vague.   maybe `await_completion` or `await_result` or `sync` is better?  i don't have strong opinion please go with what you think is best, even if that is `wait` of course.

vincbeck replied:
Agree. Done

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+
+        self.statement_id = self.execute_query()
+
+        if self.timeout > 0:
```
```suggestion
        if self.nowait is False:
```

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+
+    def wait_for_results(self, statement_id):
+        elapsed = 0
+        while elapsed < self.timeout:
```
can lean on execution_timeout
```suggestion
        while True:
```

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+            elif status == 'FAILED':
+                raise ValueError(f"RedshiftDataHook.describe_statement {status}")
+            elif status == 'ABORTED':
+                raise ValueError(f"Query {status}")
```
I would include the statement id in the message

vincbeck replied:
Done

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+    def wait_for_results(self, statement_id):
+        elapsed = 0
+        while elapsed < self.timeout:
+            self.log.info("Polling", statement_id)
```
```suggestion
            self.log.info("Polling statement %r", statement_id)
```

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+            if status == 'FINISHED':
+                return status
+            elif status == 'FAILED':
+                raise ValueError(f"RedshiftDataHook.describe_statement {status}")
```
```suggestion
                raise ValueError(f"Statement {statement_id!r} terminated with status {STATUS}.")
```

dstandish commented on airflow/providers/amazon/aws/operators/redshift_data.py:
```diff
@@ -0,0 +1,151 @@
+            elif status == 'FAILED':
+                raise ValueError(f"RedshiftDataHook.describe_statement {status}")
+            elif status == 'ABORTED':
+                raise ValueError(f"Query {status}")
```
```suggestion
                raise ValueError(f"Statement {statement_id!r} terminated with status {STATUS}.")
```

vincbeck committed:
Add redshift-data hook and operator

Use the AWS Boto3 library to query and retrieve data via the AWS API and avoid the necessity of a Postgres connection.

Co-authored-by: john-jac <jacnjoh@amazon.com>

dstandish committed:
RDS -> RedshiftData
